### PR TITLE
Add notes for n25218

### PR DESCRIPTION
**Pull request by stickies-v**
cc @furszy - feel free to have a look (no pressure). If you have any questions, comments or suggestions, happy to hear them!

### CODE REVIEW

File: _posts/2022-07-20-#25218.md [hunk 1]
```diff
@@ -7,17 +7,52 @@ authors: [furszy]
...
-commit:
+commit: 111ea3a
 ---
 
-_Notes and questions to follow soon!_
-
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+- To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful. With that pattern, one or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) can be used to access the data that was modified or created within the function, as well as to access information about how the operation happened (e.g. error messages).
```
Comment (glozow) [inline _posts/2022-07-20-#25218.md:15]:
```suggestion
- To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful, along with one or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) to access the function result and information about the operation, e.g. error messages.
```

**Review by glozow — COMMENTED** [review]
Looks good, thanks @stickes-v! Just a few nits

File: _posts/2022-07-20-#25218.md [hunk 2]
```diff
@@ -7,17 +7,52 @@ authors: [furszy]
...
+
+3. Why does [`BResult`](https://github.com/bitcoin/bitcoin/blob/111ea3ab711414236f8678566a7884d48619b2d8/src/util/result.h#L16) have a separate constructor `BResult(const bilingual_str& error)` that seems to do the exact same as the templated constructor `BResult(const T& _obj)`? Does this introduce any edge cases or limitations, and if so - are they documented?
+
+4. Do you know of any other commonly used return type(s) that are similar to `BResult`?
+
+5. In commit `wallet: refactor, include 'FeeCalculation' inside 'CreatedTransactionResult'`, what do you think is the rationale for making `FeeCalculation` a member of `CreatedTransactionResult` instead of having it as an out-parameter in `CreateTransaction()`?
+
+6. Should someone now follow up with a big PR that refactors all functions that would benefit from using `BResult`? Why (not)?
+
+7. (Bonus) Do you know of any others functions that you think would benefit nicely from being refactored to use `BResult`?
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply _posts/2022-07-20-#25218.md]: Fixed!

File: _posts/2022-07-20-#25218.md [hunk 3]
```diff
@@ -7,17 +7,52 @@ authors: [furszy]
...
+
 ## Questions
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
--->
+
+2. What are the different types of parameters and return values that the PR assumes we commonly use in a function signature? Which of these parameters are affected by the PR?
+
+3. Why does [`BResult`](https://github.com/bitcoin/bitcoin/blob/111ea3ab711414236f8678566a7884d48619b2d8/src/util/result.h#L16) have a separate constructor `BResult(const bilingual_str& error)` that seems to do the exact same as the templated constructor `BResult(const T& _obj)`? Does this introduce any edge cases or limitations, and if so - are they documented?
+
+4. Do you know of any other commonly used return type(s) that are similar to `BResult`?
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply _posts/2022-07-20-#25218.md]: Fixed!

File: _posts/2022-07-20-#25218.md [hunk 4]
```diff
@@ -7,17 +7,52 @@ authors: [furszy]
...
+    // do something...
+    if (error) {
+        error_string = "something bad happened";
+        return false;
+    }
+
+    result = goodResult;
+    return true;
+}
+```
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply _posts/2022-07-20-#25218.md:29]: Yes good idea, added the one provided in the PR

File: _posts/2022-07-20-#25218.md (hunk 1 above)
Comment (stickies-v) [reply _posts/2022-07-20-#25218.md]: Simplification here makes sense - just modified your suggestion a bit to not have "along with" to highlight that out-parameters are not being "returned".

> To simplify error handling in the [call site](https://en.wikipedia.org/wiki/Call_site), functions in Bitcoin Core quite often return a boolean to indicate if the operation was successful. One or multiple [out-parameters](http://www.cs.ecu.edu/karl/2530/spr18/Notes/lec21A.html#logical-calling-modes:~:text=an%20implicit%20one.-,Logical%20calling%20modes,-Call%20by%20value) can then be used to access the function result and information about the operation (e.g. error messages).

**Comment by stickies-v** [comment]
Thanks for the review @glozow, incorporated all your suggestions (one slightly modified).

**Comment by furszy** [comment]
Probably could mention that `BResult` internally uses an `std::variant` (which is analogous to a union). It allows multiple different types to reside in a single memory space.
So, in other words, the result class is essentially an OR: it contains the succeed object OR the failure object in-memory and not both (like before, when we were initializing the error string first and passing the reference).

**Comment by stickies-v** [comment]
Thanks @furszy, I've added an extra bullet point explaining the internals of `BResult` as per your suggestion. Will merge this already so people can start prepping but feel free to follow up with any further comments.

**Pull request merged**